lxd/storage/drivers/drivers_lvm_utils.go

```diff
@@ -0,0 +1,693 @@
+	}
+
+	volDevPath := d.lvmDevPath(vgName, vol.volType, vol.contentType, vol.name)
+	output, err = makeFSType(volDevPath, d.volumeFilesystem(vol), nil)
+	if err != nil {
+		return fmt.Errorf("Error making filesystem on LVM logical volume: %v (%s)", err, output)
+	}
+
+	d.logger.Debug("Logical volume created", log.Ctx{"vg_name": vgName, "lv_name": lvFullName, "size": fmt.Sprintf("%db", lvSizeBytes), "fs": d.volumeFilesystem(vol)})
+	return nil
+}
+
+// createLogicalVolumeSnapshot creates a snapshot of a logical volume.
+func (d *lvm) createLogicalVolumeSnapshot(vgName string, srcVol, snapVol Volume, readonly bool, makeThinLv bool) (string, error) {
+	srcVolDevPath := d.lvmDevPath(vgName, srcVol.volType, srcVol.contentType, srcVol.name)
+	isRecent, err := d.lvmVersionIsAtLeast(lvmVersion, "2.02.99")
+	if err != nil {
+		return "", fmt.Errorf("Error checking LVM version: %v", err)
+	}
+
+	snapLvName := d.lvmFullVolumeName(snapVol.volType, snapVol.contentType, snapVol.name)
+	logCtx := log.Ctx{"vg_name": vgName, "lv_name": snapLvName, "src_dev": srcVolDevPath, "thin": makeThinLv}
+	args := []string{"-n", snapLvName, "-s", srcVolDevPath}
+
+	if isRecent {
+		args = append(args, "-kn")
+	}
+
+	// If the source is not a thin volume the size needs to be specified.
+	// According to LVM tools 15-20% of the original volume should be sufficient.
+	// However, let's not be stingy at first otherwise we might force users to fiddle around with lvextend.
+	if !makeThinLv {
+		lvSizeBytes, err := d.roundedSizeBytesString(d.volumeSize(snapVol))
+		if err != nil {
+			return "", err
+		}
+
+		args = append(args, "--size", fmt.Sprintf("%db", lvSizeBytes))
+		logCtx["size"] = fmt.Sprintf("%db", lvSizeBytes)
+	}
+
+	if readonly {
+		args = append(args, "-pr")
+	} else {
+		args = append(args, "-prw")
+	}
+
+	revert := revert.New()
+	defer revert.Fail()
+
+	_, err = shared.TryRunCommand("lvcreate", args...)
+	if err != nil {
+		return "", fmt.Errorf("Error creating LVM logical volume snapshot: %s to %s: %v", srcVol.name, snapVol.name, err)
+	}
+	d.logger.Debug("Logical volume snapshot created", logCtx)
+
+	revert.Add(func() {
+		d.removeLogicalVolume(d.lvmDevPath(vgName, snapVol.volType, snapVol.contentType, snapVol.name))
+	})
+
+	targetVolDevPath := d.lvmDevPath(vgName, snapVol.volType, snapVol.contentType, snapVol.name)
+	if makeThinLv {
+		// Snapshots of thin logical volumes can be directly activated.
+		// Normal snapshots will complain about changing the origin (Which they never do.),
+		// so skip the activation since the logical volume will be automatically activated anyway.
+		err := d.activeLogicalVolume(targetVolDevPath)
+		if err != nil {
+			return "", fmt.Errorf("Error activating LVM logical volume snapshot: %s: %v", snapVol.name, err)
+		}
+	}
+
+	revert.Success()
+	return targetVolDevPath, nil
+}
+
+// activeLogicalVolume marks the logical volume as active.
+func (d *lvm) activeLogicalVolume(volDevPath string) error {
```

Comment:
Only has a single caller, any reason to keep it separate?

Reply:
No will merge into createLogicalVolumeSnapshot